pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go

```diff
@@ -208,6 +209,27 @@ func (pl *nonCSILimits) EventsToRegister() []framework.ClusterEvent {
 	}
 }
 
+// PreFilter invoked at the prefilter extension point
+//
+// If the pod haven't those types of volumes, we'll skip the Filter phase
+func (pl *nonCSILimits) PreFilter(ctx context.Context, _ *framework.CycleState, pod *v1.Pod) (*framework.PreFilterResult, *framework.Status) {
+	volumes := pod.Spec.Volumes
+	for i := range volumes {
+		vol := &volumes[i]
+		_, ok := pl.filter.FilterVolume(vol)
+		if ok || vol.PersistentVolumeClaim != nil || vol.Ephemeral != nil {
```

Comment:
> shouldn't we check that the PVC passes `matchProvisioner`?

I think `vol.PersistentVolumeClaim != nil || vol.Ephemeral != nil` has already include `matchProvisioner`, should we check more?

https://github.com/kubernetes/kubernetes/blob/5ede346fc8c1558e4d3e67aa0d96da68eb1f727e/pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go#L308-L325

Reply:
> shouldn't we check that the PVC passes `matchProvisioner`?

Yes and no.

I think the current state is correct: if the condition "ok || vol.PersistentVolumeClaim != nil || vol.Ephemeral != nil" is true, this mean that filterVolumes **may** include a given volume in a "filteredVolumes" set so we cannot short cut. If all volumes are not matching the condition, then we know that for sure filteredVolumes will return empty result.

More specifically if vol.PersistentVolumeClaim != nil, but NOT `pl.matchProvisioner`, we will not return Skip in PreFilter and in Filter we will not add element in filterVolumes, which is correct, but is a potential waste of time.

Reply:
Is it feasible actually here to run filterVolumes directly and:
* store result in plugin's state so that we won't recompute this in context of all nodes (may save some CPU)
* short cut (return Skip) if we have an empty result?

This way we both save CPU and reduce risk of inaccuracy between PreFilter conditions and actual filterVolumes implementation.

Performance-wise I think it should be similar:
* if all volumes are not meeting "ok || vol.PersistentVolumeClaim != nil || vol.Ephemeral != nil" condition, we will just "continue" them in filterVolumes

Reply:
> Sounds good. Let's leave it for a follow up.

Thanks, I'd like to follow up.